Improve c2-aten tensor interop testing

dzhulgakov opened the pull request:
Summary:
Few changes (need to split into multiple diffs probably):
- make conversion explicit (as they can throw to avoid surprises)
- tensor dispatch is actually not initialized - ezyang - can you suggest where to plug proper fix (what I did is inefficient but does the job)
- there are more invariants we might need to enforce (e.g. non-POD types), adding a place to verify it in enforce_invariants

Needs verification - invoking enforce_invariants might be expensive because this constructor is invoked in IValue all the time, potentially a better place needs to be found

Differential Revision: D13596031

dzhulgakov commented on caffe2/core/tensor.cc:
```diff
@@ -175,6 +175,15 @@ void ReinitializeAndCopyFrom(
+  }
+  CAFFE_ENFORCE(
+      impl_->is_contiguous(),
+      "Caffe2 tensor wrapper supports only contiguous tensors");
```
Yes, I'm aware of this problem (and will add it to comments). I agree with you that the cure is worse than disease. Wrapping tensors around should be short-lived or taken with care. So I guess it's ok to have it like this

li-roy commented:
@ezyang @dzhulgakov Actually the fix I ended up using in thpp was to initialize in the constructor of a static object like so:

```
class ContextInitializer {
 public:
  ContextInitializer() {
    at::init();
  }
};

static ContextInitializer contextInitializer_;
```

Commit:
Improve c2-aten tensor interop and add proper testing (#15860)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15860

Few changes (which are harder to split in separate diffs, so together):
- make conversion explicit (as they can throw to avoid surprises)
- fix tensor legacy dispatch not initialized when tensor is created on C2 side
- add a bunch of invariants to enforce

Differential Revision: D13596031

fbshipit-source-id: ab08d6ba0a4f14210e6eb0bbfba6881b17e614d5
